File: crates/ibex2/src/bindings/intl_number_format.js
// ECMA-402 2020 NumberFormat object plumbing for Linux Hermes.
// Observable JavaScript coercion order stays here; all normalized policy and
// formatting semantics live in Rust and ICU through the private binding.
(function (global) {
  "use strict";

  var raw = global.__ibex2_intl_number_format;
  delete global.__ibex2_intl_number_format;

  var IntlObject = global.Intl;
  var canonicalLocales = IntlObject.getCanonicalLocales;
  var reflectApply = Reflect.apply;
  var objectCreate = Object.create;
  var objectDefineProperty = Object.defineProperty;
  var objectDefineProperties = Object.defineProperties;
  var numberValueOf = Number.prototype.valueOf;
  var bigintValueOf = typeof BigInt === "function" ? BigInt.prototype.valueOf : null;
  var bigintToString = typeof BigInt === "function" ? BigInt.prototype.toString : null;
  var stringConcat = String.prototype.concat;
  var floor = Math.floor;
  var boundFormats = new WeakMap();

  var simpleUnits = Object.create(null);
  var unitNames = [
    "acre", "bit", "byte", "celsius", "centimeter", "day", "degree",
    "fahrenheit", "fluid-ounce", "foot", "gallon", "gigabit", "gigabyte",
    "gram", "hectare", "hour", "inch", "kilobit", "kilobyte", "kilogram",
    "kilometer", "liter", "megabit", "megabyte", "meter", "mile",
    "mile-scandinavian", "milliliter", "millimeter", "millisecond", "minute",
    "month", "ounce", "percent", "petabyte", "pound", "second", "stone",
    "terabit", "terabyte", "week", "yard", "year"
  ];
  for (var unitIndex = 0; unitIndex < unitNames.length; unitIndex++) {
    simpleUnits[unitNames[unitIndex]] = true;
  }

  function optionObject(options) {
    if (options === undefined) return objectCreate(null);
    if (options === null) throw new TypeError("options must not be null");
    return Object(options);
  }

  function stringOption(options, name, allowed, fallback) {
    var value = options[name];
    if (value === undefined) return fallback;
    // Unlike the String constructor, abstract ToString rejects Symbols.
    value = reflectApply(stringConcat, "", [value]);
    if (allowed !== undefined && allowed.indexOf(value) < 0) {
      throw new RangeError(name + " is not a supported value");
    }
    return value;
  }

  function defaultNumberOption(value, minimum, maximum, fallback, name) {
    if (value === undefined) return fallback;
    // Unary plus is ToNumber: unlike the Number constructor, it rejects both
    // primitive and boxed BigInt values.
    value = +value;
    if (value !== value || value < minimum || value > maximum) {
      throw new RangeError(name + " is outside its allowed range");
    }
    return floor(value);
  }

  function validUnicodeType(value) {
    var parts = value.split("-");
    for (var i = 0; i < parts.length; i++) {
      if (!/^[0-9A-Za-z]{3,8}$/.test(parts[i])) return false;
    }
    return true;
  }

  function validCurrency(value) {
    return /^[A-Za-z]{3}$/.test(value);
  }

  function asciiUpper(value) {
    return value.replace(/[a-z]/g, function (letter) {
      return String.fromCharCode(letter.charCodeAt(0) - 32);
    });
  }

  function validUnit(value) {
    if (simpleUnits[value]) return true;
    var marker = "-per-";
    var first = value.indexOf(marker);
    return first > 0 && first === value.lastIndexOf(marker) &&
      simpleUnits[value.slice(0, first)] === true &&
      simpleUnits[value.slice(first + marker.length)] === true;
  }

  function canonicalize(locales) {
    return reflectApply(canonicalLocales, IntlObject, [locales]);
  }

  function normalize(locales, options) {
    var requested = canonicalize(locales);
    var opts = optionObject(options);
    var localeMatcher = stringOption(opts, "localeMatcher", ["lookup", "best fit"], "best fit");
    var numberingSystem = stringOption(opts, "numberingSystem", undefined, undefined);
    if (numberingSystem !== undefined && !validUnicodeType(numberingSystem)) {
      throw new RangeError("numberingSystem is not a Unicode locale type");
    }
    var style = stringOption(opts, "style", ["decimal", "percent", "currency", "unit"], "decimal");
    var currency = stringOption(opts, "currency", undefined, undefined);
    if (currency !== undefined && !validCurrency(currency)) {
      throw new RangeError("currency is not a well-formed currency code");
    }
    if (style === "currency" && currency === undefined) {
      throw new TypeError("currency is required with currency style");
    }
    var currencyDisplay = stringOption(opts, "currencyDisplay", ["code", "symbol", "narrowSymbol", "name"], "symbol");
    var currencySign = stringOption(opts, "currencySign", ["standard", "accounting"], "standard");
    var unit = stringOption(opts, "unit", undefined, undefined);
    if (unit !== undefined && !validUnit(unit)) {
      throw new RangeError("unit is not a sanctioned unit identifier");
    }
    if (style === "unit" && unit === undefined) {
      throw new TypeError("unit is required with unit style");
    }
    var unitDisplay = stringOption(opts, "unitDisplay", ["short", "narrow", "long"], "short");
    if (currency !== undefined) currency = asciiUpper(currency);

    var minimumFractionDefault = style === "currency" ? raw.currencyDigits(currency) : 0;
    var maximumFractionDefault = style === "currency" ? minimumFractionDefault :
      (style === "percent" ? 0 : 3);

    var notation = stringOption(opts, "notation", ["standard", "scientific", "engineering", "compact"], "standard");
    var minimumIntegerDigits = defaultNumberOption(opts.minimumIntegerDigits, 1, 21, 1, "minimumIntegerDigits");
    // ECMA-402 2020 reads all four properties in this order, then coerces only
    // the significant pair or the fraction pair selected by precedence.
    var minimumFractionDigits = opts.minimumFractionDigits;
    var maximumFractionDigits = opts.maximumFractionDigits;
    var minimumSignificantDigits = opts.minimumSignificantDigits;
    var maximumSignificantDigits = opts.maximumSignificantDigits;
    if (minimumSignificantDigits !== undefined || maximumSignificantDigits !== undefined) {
      minimumSignificantDigits = defaultNumberOption(minimumSignificantDigits, 1, 21, 1, "minimumSignificantDigits");
      maximumSignificantDigits = defaultNumberOption(maximumSignificantDigits, minimumSignificantDigits, 21, 21, "maximumSignificantDigits");
      minimumFractionDigits = undefined;
      maximumFractionDigits = undefined;
    } else if (minimumFractionDigits !== undefined || maximumFractionDigits !== undefined) {
      minimumFractionDigits = defaultNumberOption(minimumFractionDigits, 0, 20, minimumFractionDefault, "minimumFractionDigits");
      var actualMaximumDefault = Math.max(minimumFractionDigits, maximumFractionDefault);
      maximumFractionDigits = defaultNumberOption(maximumFractionDigits, minimumFractionDigits, 20, actualMaximumDefault, "maximumFractionDigits");
    } else if (notation !== "compact") {
      minimumFractionDigits = minimumFractionDefault;
      maximumFractionDigits = maximumFractionDefault;
    } else {
      minimumFractionDigits = undefined;
      maximumFractionDigits = undefined;
    }
    var compactDisplay = stringOption(opts, "compactDisplay", ["short", "long"], "short");
    var useGroupingValue = opts.useGrouping;
    var useGrouping = useGroupingValue === undefined ? true : Boolean(useGroupingValue);
    var signDisplay = stringOption(opts, "signDisplay", ["auto", "never", "always", "exceptZero"], "auto");

    return [requested.join("\n"), localeMatcher, numberingSystem, style,
      currency, currencyDisplay, currencySign, unit, unitDisplay, notation,
      minimumIntegerDigits, minimumFractionDigits, maximumFractionDigits,
      minimumSignificantDigits, maximumSignificantDigits, compactDisplay,
      useGrouping, signDisplay];
  }

  function toPrimitiveNumber(value) {
    if (value === null || (typeof value !== "object" && typeof value !== "function")) {
      return value;
    }
    var exotic = value[Symbol.toPrimitive];
    if (exotic !== undefined && exotic !== null) {
      if (typeof exotic !== "function") throw new TypeError("Symbol.toPrimitive is not callable");
      var result = reflectApply(exotic, value, ["number"]);
      if (result === null || (typeof result !== "object" && typeof result !== "function")) return result;
      throw new TypeError("cannot convert object to primitive value");
    }
    for (var i = 0; i < 2; i++) {
      var method = value[i === 0 ? "valueOf" : "toString"];
      if (typeof method === "function") {
        var primitive = reflectApply(method, value, []);
        if (primitive === null || (typeof primitive !== "object" && typeof primitive !== "function")) return primitive;
      }
    }
    throw new TypeError("cannot convert object to primitive value");
  }

  function numeric(value) {
    var primitive = toPrimitiveNumber(value);
    if (typeof primitive === "bigint") {
      return ["bigint", reflectApply(bigintToString, primitive, [])];
    }
    return ["number", Number(primitive)];
  }

  function NumberFormat(locales, options) {
    if (new.target === undefined) return new NumberFormat(locales, options);
    var normalized = normalize(locales, options);
    raw.initialize.apply(raw, [this].concat(normalized));
  }

  function getFormat() {
    raw.assertReceiver(this);
    var bound = boundFormats.get(this);
    if (bound === undefined) {
      var receiver = this;
      bound = function format(value) {
        var input = numeric(value);
        return raw.format(receiver, input[0], input[1]);
      };
      boundFormats.set(this, bound);
    }
    return bound;
  }

  function formatToParts(value) {
    raw.assertReceiver(this);
    var input = numeric(value);
    return raw.formatToParts(this, input[0], input[1]);
  }

  function resolvedOptions() {
    return raw.resolvedOptions(this);
  }

  function supportedLocalesOf(locales, options) {
    var requested = canonicalize(locales);
    var opts = optionObject(options);
    var matcher = stringOption(opts, "localeMatcher", ["lookup", "best fit"], "best fit");
    var supported = raw.supportedLocalesOf(requested.join("\n"), matcher);
    return supported === "" ? [] : supported.split("\n");
  }

  objectDefineProperties(NumberFormat.prototype, {
    constructor: { value: NumberFormat, writable: true, configurable: true },
    format: { get: getFormat, configurable: true },
    formatToParts: { value: formatToParts, writable: true, configurable: true },
    resolvedOptions: { value: resolvedOptions, writable: true, configurable: true }
  });
  objectDefineProperty(NumberFormat.prototype, Symbol.toStringTag, {
    value: "Intl.NumberFormat", configurable: true
  });
  objectDefineProperty(NumberFormat, "supportedLocalesOf", {
    value: supportedLocalesOf, writable: true, configurable: true
  });
  objectDefineProperty(IntlObject, "NumberFormat", {
    value: NumberFormat, writable: true, configurable: true
  });

  function numberToLocaleString() {
    var value = reflectApply(numberValueOf, this, []);
    return new NumberFormat(arguments[0], arguments[1]).format(value);
  }
  objectDefineProperty(Number.prototype, "toLocaleString", {
    value: numberToLocaleString, writable: true, configurable: true
  });

  if (bigintValueOf !== null) {
    function bigintToLocaleString() {
      var value = reflectApply(bigintValueOf, this, []);
      return new NumberFormat(arguments[0], arguments[1]).format(value);
    }
    objectDefineProperty(BigInt.prototype, "toLocaleString", {
      value: bigintToLocaleString, writable: true, configurable: true
    });
  }
})(globalThis);

File: crates/ibex2/tests/intl_number_format.rs
//! Native ICU-backed NumberFormat behavior on the Linux vanilla-Hermes profile.
#![cfg(all(feature = "hermes", target_os = "linux"))]

use ibex2::engine::hermes::{DynamicCode, Hermes};

fn runtime(hardened: bool) -> Hermes {
    let mut runtime = Hermes::new(DynamicCode::Closed).expect("runtime");
    assert!(runtime.install_stdlib());
    runtime.install_bindings().expect("bindings");
    if hardened {
        runtime.harden().expect("harden");
    }
    runtime
}

fn eval(runtime: &mut Hermes, source: &str) -> String {
    runtime
        .eval(source)
        .unwrap_or_else(|error| panic!("{source}: {error}"))
}

#[test]
fn formats_locales_styles_precision_and_numbering_systems() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              function f(locale, options, value) {
                return new Intl.NumberFormat(locale, options).format(value);
              }
              var grouped = f('de-DE', {}, 1234.5);
              var currency = f('en-US', {style:'currency', currency:'USD'}, 1234.5);
              var percent = f('en-US', {style:'percent'}, 0.56);
              var unit = f('en-US', {style:'unit', unit:'kilometer-per-hour'}, 80);
              var scientific = f('en-US', {notation:'scientific'}, 12345);
              var engineering = f('en-US', {notation:'engineering'}, 12345);
              var compact = f('en-US', {notation:'compact'}, 1200);
              var significant = f('en-US', {minimumSignificantDigits:4, maximumSignificantDigits:4}, 12.3);
              var arab = f('ar-EG', {numberingSystem:'arab', useGrouping:false}, 123);
              return [
                grouped.indexOf('1.234') >= 0 && grouped.indexOf(',5') >= 0,
                currency.indexOf('$') >= 0 && currency.indexOf('1,234.50') >= 0,
                percent === '56%',
                unit.indexOf('80') >= 0 && /km|kilometer/.test(unit),
                /E4/.test(scientific), /E3/.test(engineering), compact === '1.2K',
                significant.indexOf('12.30') >= 0,
                arab !== '123' && arab.length === 3
              ].join(':');
            })()"#,
        ),
        "true:true:true:true:true:true:true:true:true"
    );
}

#[test]
fn parts_join_format_and_classify_special_values_and_signs() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var nf = new Intl.NumberFormat('en-US', {
                style:'currency', currency:'USD', signDisplay:'always'
              });
              function check(value, expected) {
                var parts = nf.formatToParts(value);
                var joined = parts.map(function (part) { return part.value; }).join('');
                var types = parts.map(function (part) { return part.type; });
                return joined === nf.format(value) && expected.every(function (type) {
                  return types.indexOf(type) >= 0;
                });
              }
              var decimal = new Intl.NumberFormat('en-US', {minimumFractionDigits:2});
              var fractionParts = decimal.formatToParts(1234.5);
              var fractionTypes = fractionParts.map(function (part) { return part.type; });
              return [
                check(12, ['plusSign','currency','integer']),
                check(-0, ['minusSign','currency','integer']),
                check(NaN, ['plusSign','currency','nan']),
                check(Infinity, ['plusSign','currency','infinity']),
                fractionParts.map(function (part) { return part.value; }).join('') === decimal.format(1234.5),
                ['integer','group','decimal','fraction'].every(function (type) {
                  return fractionTypes.indexOf(type) >= 0;
                })
              ].join(':');
            })()"#,
        ),
        "true:true:true:true:true:true"
    );
}

#[test]
fn normalizes_nan_sign_and_preserves_real_signed_values() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              function show(options, value) {
                var nf = new Intl.NumberFormat('en-US', options);
                return nf.format(value) + '=' + nf.formatToParts(value).map(function (part) {
                  return part.type + '(' + part.value + ')';
                }).join(',');
              }
              return [
                show({}, NaN), show({signDisplay:'always'}, NaN),
                show({style:'currency',currency:'USD'}, NaN),
                show({style:'currency',currency:'USD',signDisplay:'always'}, NaN),
                show({}, -0), show({}, 0), show({signDisplay:'always'}, 0)
              ].join('|');
            })()"#,
        ),
        concat!(
            "NaN=nan(NaN)|+NaN=plusSign(+),nan(NaN)|",
            "$NaN=currency($),nan(NaN)|+$NaN=plusSign(+),currency($),nan(NaN)|",
            "-0=minusSign(-),integer(0)|0=integer(0)|+0=plusSign(+),integer(0)"
        )
    );
}

#[test]
fn sign_display_matrix_includes_special_and_rounded_zero_values() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var values = [NaN, -0, 0, -1, 1, -Infinity, Infinity, -0.001, 0.001];
              return ['auto','never','always','exceptZero'].map(function (mode) {
                var nf = new Intl.NumberFormat('en-US', {
                  signDisplay:mode, maximumFractionDigits:0, useGrouping:false
                });
                return mode + '=' + values.map(nf.format).join('|');
              }).join(';');
            })()"#,
        ),
        concat!(
            "auto=NaN|-0|0|-1|1|-∞|∞|-0|0;",
            "never=NaN|0|0|1|1|∞|∞|0|0;",
            "always=+NaN|-0|+0|-1|+1|-∞|+∞|-0|+0;",
            "exceptZero=NaN|0|0|-1|+1|-∞|+∞|0|0"
        )
    );
}

#[test]
fn resolved_options_and_locale_negotiation_match_selected_configuration() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var same = new Intl.NumberFormat('de-u-nu-arab', {numberingSystem:'arab'}).resolvedOptions();
              var changed = new Intl.NumberFormat('de-u-nu-arab', {numberingSystem:'latn'}).resolvedOptions();
              var privateUse = new Intl.NumberFormat('en-x-u-nu-arab').resolvedOptions();
              var upperOption = new Intl.NumberFormat('en', {numberingSystem:'ARAB'}).resolvedOptions();
              var upperSame = new Intl.NumberFormat('de-u-nu-arab', {numberingSystem:'ARAB'}).resolvedOptions();
              var unsupportedOverride = new Intl.NumberFormat(
                'de-u-nu-arab', {numberingSystem:'foobar'}).resolvedOptions();
              var unsupportedType = new Intl.NumberFormat('en-u-nu-arab-foobar').resolvedOptions();
              var currency = new Intl.NumberFormat(['zz-ZZ','de-DE'], {
                style:'currency', currency:'KWD', currencyDisplay:'code',
                currencySign:'accounting', minimumIntegerDigits:2,
                minimumFractionDigits:3, maximumFractionDigits:4,
                useGrouping:false, signDisplay:'exceptZero'
              }).resolvedOptions();
              var supported = Intl.NumberFormat.supportedLocalesOf(['de-DE','zz-ZZ']);
              return [
                same.locale.indexOf('-u-nu-arab') >= 0, same.numberingSystem === 'arab',
                changed.locale.indexOf('-u-nu-') < 0, changed.numberingSystem === 'latn',
                privateUse.numberingSystem === 'latn' && privateUse.locale.indexOf('-u-nu-') < 0,
                upperOption.numberingSystem === 'arab',
                upperSame.numberingSystem === 'arab' && upperSame.locale.indexOf('-u-nu-arab') >= 0,
                unsupportedOverride.numberingSystem === 'arab',
                unsupportedOverride.locale.indexOf('-u-nu-arab') >= 0,
                unsupportedType.numberingSystem === 'latn' && unsupportedType.locale.indexOf('-u-nu-') < 0,
                currency.locale.indexOf('de') === 0, currency.style === 'currency',
                currency.currency === 'KWD', currency.currencyDisplay === 'code',
                currency.currencySign === 'accounting', currency.minimumIntegerDigits === 2,
                currency.minimumFractionDigits === 3, currency.maximumFractionDigits === 4,
                currency.useGrouping === false, currency.signDisplay === 'exceptZero',
                supported.length === 1 && supported[0] === 'de-DE'
              ].join(':');
            })()"#,
        ),
        "true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true:true"
    );
}

#[test]
fn coercion_order_and_ecmascript_conversion_rules_are_preserved() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var results = [];
              function errorName(thunk) {
                try { thunk(); return 'accepted'; } catch (error) { return error.name; }
              }
              results.push(errorName(function () {
                new Intl.NumberFormat('en', {style:Symbol('decimal')});
              }));
              results.push(errorName(function () {
                new Intl.NumberFormat('en', {minimumIntegerDigits:1n});
              }));
              var nullable = {
                [Symbol.toPrimitive]: null,
                valueOf: function () { return 4; }
              };
              results.push(new Intl.NumberFormat('en', {minimumIntegerDigits:nullable})
                .resolvedOptions().minimumIntegerDigits);

              var order = [];
              var fraction = { valueOf: function () { order.push('coerce fraction'); return 2; } };
              var options = {
                get minimumFractionDigits() { order.push('get min fraction'); return fraction; },
                get maximumFractionDigits() { order.push('get max fraction'); return fraction; },
                get minimumSignificantDigits() { order.push('get min significant'); return 3; },
                get maximumSignificantDigits() { order.push('get max significant'); return 5; }
              };
              new Intl.NumberFormat('en', options);
              results.push(order.join(','));

              var touched = false;
              var value = { valueOf: function () { touched = true; return 1; } };
              results.push(errorName(function () {
                Intl.NumberFormat.prototype.formatToParts.call({}, value);
              }) + ':' + touched);
              return results.join('|');
            })()"#,
        ),
        "TypeError|TypeError|4|get min fraction,get max fraction,get min significant,get max significant|TypeError:false"
    );
}

#[test]
fn construction_uses_new_target_and_never_symbol_has_instance() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var NF = Intl.NumberFormat;
              var log = [], customPrototype = Object.create(NF.prototype);
              function DerivedTarget() {}
              var Derived = new Proxy(DerivedTarget, {get:function (target, name, receiver) {
                if (name === 'prototype') { log.push('prototype'); return customPrototype; }
                return Reflect.get(target, name, receiver);
              }});
              var locales = {length:1, get 0() { log.push('locale'); return 'en-US'; }};
              var derived = Reflect.construct(NF, [locales], Derived);
              var samePrototype = Object.getPrototypeOf(derived) === customPrototype;
              var works = Object.getOwnPropertyDescriptor(NF.prototype, 'format').get.call(derived)(12) === '12';
              Object.defineProperty(NF, Symbol.hasInstance, {
                value: function () { throw new Error('consulted'); }, configurable: true
              });
              var ordinary = NF('en-US');
              return [samePrototype, works, log.join(','),
                ordinary.resolvedOptions().locale.indexOf('en') === 0].join(':');
            })()"#,
        ),
        "true:true:prototype,locale:true"
    );
}

#[test]
#[ignore = "pinned Hermes cannot express the built-in constructor fallback through public JSI; tracked in the Linux Intl issue"]
fn non_object_new_target_prototype_uses_the_intl_prototype_fallback() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              function PrimitiveTarget() {}
              PrimitiveTarget.prototype = 0;
              var fallback = Reflect.construct(Intl.NumberFormat, ['en'], PrimitiveTarget);
              return Object.getPrototypeOf(fallback) === Intl.NumberFormat.prototype;
            })()"#,
        ),
        "true"
    );
}

#[test]
fn number_and_bigint_locale_methods_share_exact_native_formatting() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var options = {useGrouping:false, maximumFractionDigits:2};
              var nf = new Intl.NumberFormat('en-US', options);
              var huge = 123456789012345678901234567890n;
              return [
                (1234.5).toLocaleString('en-US', options) === nf.format(1234.5),
                huge.toLocaleString('en-US', {useGrouping:false}) === '123456789012345678901234567890',
                huge.toLocaleString('de-DE') === new Intl.NumberFormat('de-DE').format(huge)
              ].join(':');
            })()"#,
        ),
        "true:true:true"
    );
}

#[test]
fn bound_format_retains_native_state_without_exposing_handles() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var formatter = new Intl.NumberFormat('de-DE', {minimumFractionDigits:2});
              globalThis.savedIntlFormat = formatter.format;
              return [formatter.format === formatter.format,
                Object.getOwnPropertyNames(formatter).length,
                typeof globalThis.__ibex2_intl_number_format].join(':');
            })()"#,
        ),
        "true:0:undefined"
    );
    assert!(runtime.collect_garbage());
    assert_eq!(
        eval(
            &mut runtime,
            "savedIntlFormat(1234.5).indexOf('1.234,50') >= 0 ? 'alive' : 'wrong'",
        ),
        "alive"
    );
}

#[test]
fn public_methods_have_builtin_names_lengths_and_are_not_constructors() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var nf = new Intl.NumberFormat('en-US');
              var getter = Object.getOwnPropertyDescriptor(
                Intl.NumberFormat.prototype, 'format').get;
              var methods = [Intl.NumberFormat.supportedLocalesOf, getter, nf.format,
                Intl.NumberFormat.prototype.formatToParts,
                Intl.NumberFormat.prototype.resolvedOptions,
                Number.prototype.toLocaleString, BigInt.prototype.toLocaleString];
              function rejectsConstruction(fn) {
                try { Reflect.construct(fn, []); return false; }
                catch (error) { return error.name === 'TypeError'; }
              }
              return [
                methods.every(rejectsConstruction),
                methods.map(function (fn) { return fn.name; }).join(','),
                methods.map(function (fn) { return fn.length; }).join(',')
              ].join('|');
            })()"#,
        ),
        "true|supportedLocalesOf,get format,,formatToParts,resolvedOptions,toLocaleString,toLocaleString|1,0,1,1,0,0,0"
    );
}

#[test]
fn invalid_options_throw_the_required_error_kinds() {
    let mut runtime = runtime(false);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              function kind(options) {
                try { new Intl.NumberFormat('en', options); return 'accepted'; }
                catch (error) { return error.name; }
              }
              return [
                kind({style:'currency'}), kind({style:'currency',currency:'US'}),
                kind({style:'unit'}), kind({style:'unit',unit:'horses'}),
                kind({numberingSystem:'ab'}), kind({minimumIntegerDigits:0}),
                kind({minimumFractionDigits:4,maximumFractionDigits:2}),
                kind({minimumSignificantDigits:4,maximumSignificantDigits:2}),
                kind({notation:'unknown'}), kind({signDisplay:'sometimes'})
              ].join(':');
            })()"#,
        ),
        "TypeError:RangeError:TypeError:RangeError:RangeError:RangeError:RangeError:RangeError:RangeError:RangeError"
    );
}

#[test]
fn hardening_preserves_public_surface_and_removes_private_factory() {
    let mut runtime = runtime(true);
    assert_eq!(
        eval(
            &mut runtime,
            r#"(function () {
              var nf = new Intl.NumberFormat('en-US');
              return [
                typeof globalThis.__ibex2_intl_number_format,
                Object.isFrozen(Intl.NumberFormat.prototype),
                Intl.NumberFormat.length,
                Intl.NumberFormat.supportedLocalesOf.length,
                Object.getOwnPropertyDescriptor(Intl.NumberFormat, 'prototype').writable,
                Object.getOwnPropertyDescriptor(Intl.NumberFormat.prototype, 'format').get.name,
                JSON.stringify(nf.format.name), nf.format.length,
                Intl.NumberFormat.prototype.constructor === Intl.NumberFormat,
                Object.prototype.toString.call(nf),
                nf.format(1234)
              ].join(':');
            })()"#,
        ),
        "undefined:true:0:1:false:get format:\"\":1:true:[object Intl.NumberFormat]:1,234"
    );
}
